### DISCUSSION TITLE
Feature selection package based on SHAP and target permutation

### DISCUSSION THREAD

**Discussion by manuel-calzolari** [shap/shap#2577] [Show and tell]
I have developed an experimental feature selection library based on SHAP: https://github.com/manuel-calzolari/shapicant

It is inspired by this article: https://academic.oup.com/bioinformatics/article/26/10/1340/193348

Basically, it trains a model with the true target column and computes the feature importance. Then it shuffles the target, trains a new model and recomputes the feature importance. This second step is performed multiple times. The idea is that when a feature is relevant, its feature importance when the model is trained with the true target will be consistently higher than its importance when the model is trained with shuffled targets.

The original article used the Gini importance as a feature importance measure. Instead, I use SHAP. However, instead of taking the mean absolute value of the SHAP values ​​for each feature as feature importance, I take the mean value for positive and negative SHAP values ​​separately (no `abs`). The true importance (with the true target) needs to be consistently higher than null importances (with shuffled targets) for both positive and negative impacts at the same time (for multi-class classification, the true importance needs to be higher for at least one of the classes).

The idea behind taking the mean value for positive and negative SHAP values ​​separately (without `abs`) is mainly inspired by issue 553: it looks like you can get SHAP values ​​all with the same sign if the distribution of a feature is very different between the training set and the "validation set" (hold out set) that you can use to compute the SHAP values.

In that case, if I take `mean(abs(shap_values))`, I would probably have a very important feature because the mean absolute value is high. Still, that feature is not really relevant for the model to generalize well, because its distribution on the validation set is very different. However, if I consider positive and negative SHAP values ​​separately, I would probably remove that feature because its positive SHAP values ​​are zero (all the SHAP values ​​for that feature are negative).

So basically, by considering positive and negative SHAP values ​​separately, I can also discard features such as features that drift heavily and change distribution over time (between the training set and a "validation set").

In your opinion, does what I do make sense? Thank you very much!
